lib/concurrent/atomic/atomic_markable_reference.rb

```diff
@@ -0,0 +1,147 @@
+module Concurrent
+  module Edge
+    # @!macro atomic_markable_reference
+    class AtomicMarkableReference < ::Concurrent::Synchronization::Object
+      # @!macro [attach] atomic_markable_reference_method_initialize
+      def initialize(value = nil, mark = false)
+        super
+        @Reference = AtomicReference.new ImmutableArray[value, mark]
+        ensure_ivar_visibility!
+      end
+
+      # @!macro [attach] atomic_markable_reference_method_compare_and_set
+      #
+      #   Atomically sets the value and mark to the given updated value and
+      #   mark given both:
+      #     - the current value == the expected value &&
+      #     - the current mark == the expected mark
+      #
+      #   @param [Object] old_val the expected value
+      #   @param [Object] new_val the new value
+      #   @param [Boolean] old_mark the expected mark
+      #   @param [Boolean] new_mark the new mark
+      #
+      #   @return [Boolean] `true` if successful. A `false` return indicates
+      #   that the actual value was not equal to the expected value or the
+      #   actual mark was not equal to the expected mark
+      def compare_and_set(expected_val, new_val, expected_mark, new_mark)
+        # Memoize a valid reference to the current AtomicReference for
+        # later comparison.
+        current = @Reference.get
+        curr_val, curr_mark = current
+
+        # Ensure that that the expected values match.
+        return false unless (expected_val == curr_val) &&
+                            (expected_mark == curr_mark)
+
+        # In this case, it would be redundant to set the fields. Just
+        # shortcircuit without wasting CPU time on CAS.
+        return true if (new_val == curr_val) &&
+                       (new_mark == curr_mark)
```

Comment:
Yeah, these should be `equal?`/`eql?` calls for the correct semantics.

I also think this whole optimistic `if` branch should be deleted, it assumes a pretty weird call and I also expect `cas` call to only ever return `true` if my thread actually performed and succeeded in a native CAS instruction.